File: frontend/src/lib/loans.ts
export type LoanType = 'immo' | 'conso' | 'auto' | 'other';

export type AmountSource = 'calendar' | 'bank';

export interface Loan {
  id: number;
  name: string;
  loan_type: LoanType;
  initial_capital: number;
  monthly_payment: number;
  total_months: number;
  start_date: string;
  archived: boolean;
  created_at: string;
  end_date: string;
  months_paid: number;
  months_remaining: number;
  amount_remaining: number;
  progress_pct: number;
  is_active: boolean;
  linked_account_id: string | null;
  linked_label: string | null;
  amount_source: AmountSource;
}

export interface LoanCandidate {
  account_id: string;
  label: string;
  balance: number;
  currency: string;
  connector_type: string;
  as_of: string | null;
}

export interface LoanCreatePayload {
  name: string;
  loan_type: LoanType;
  initial_capital: number;
  monthly_payment: number;
  total_months: number;
  start_date: string;
}

export interface LoanSummary {
  total_monthly_payment: number;
  total_amount_remaining: number;
  last_end_date: string | null;
  active_count: number;
}

export const LOAN_TYPE_LABELS: Record<LoanType, string> = {
  immo: 'Immobilier',
  conso: 'Consommation',
  auto: 'Auto',
  other: 'Autre',
};

File: src/schemas/loans.py
from datetime import datetime
from typing import Literal
from pydantic import BaseModel, Field


LoanType = Literal["immo", "conso", "auto", "other"]


class LoanBase(BaseModel):
    name: str
    loan_type: LoanType
    initial_capital: float = Field(gt=0)
    monthly_payment: float = Field(gt=0)
    total_months: int = Field(gt=0)
    start_date: str  # ISO YYYY-MM-DD


class LoanCreate(LoanBase):
    pass


class LoanUpdate(BaseModel):
    name: str | None = None
    loan_type: LoanType | None = None
    initial_capital: float | None = Field(default=None, gt=0)
    monthly_payment: float | None = Field(default=None, gt=0)
    total_months: int | None = Field(default=None, gt=0)
    start_date: str | None = None
    archived: bool | None = None


class LoanResponse(LoanBase):
    id: int
    archived: bool
    created_at: str
    # Champs calculés :
    end_date: str
    months_paid: int
    months_remaining: int
    amount_remaining: float
    progress_pct: float
    is_active: bool


class LoanSummary(BaseModel):
    total_monthly_payment: float
    total_amount_remaining: float
    last_end_date: str | None
    active_count: int


class LoanCandidate(BaseModel):
    account_id: str
    label: str
    balance: float
    currency: str
    connector_type: str
    as_of: datetime | None = None


class LinkRequest(BaseModel):
    account_id: str
